storage-proofs/src/post/fallback/compound.rs

```diff
@@ -0,0 +1,361 @@
+use std::marker::PhantomData;
+
+use anyhow::{anyhow, ensure};
+use bellperson::Circuit;
+use fil_sapling_crypto::jubjub::JubjubEngine;
+use generic_array::typenum;
+use paired::bls12_381::{Bls12, Fr};
+
+use crate::compound_proof::{CircuitComponent, CompoundProof};
+use crate::error::Result;
+use crate::gadgets::por::PoRCompound;
+use crate::hasher::Hasher;
+use crate::parameter_cache::{CacheableParameters, ParameterSetMetadata};
+use crate::por;
+use crate::post::fallback::{self, FallbackPoSt, FallbackPoStCircuit};
+use crate::proof::ProofScheme;
+use crate::util::NODE_SIZE;
+
+use super::circuit::Sector;
+
+pub struct FallbackPoStCompound<H>
+where
+    H: Hasher,
+{
+    _h: PhantomData<H>,
+}
+
+impl<E: JubjubEngine, C: Circuit<E>, P: ParameterSetMetadata, H: Hasher>
+    CacheableParameters<E, C, P> for FallbackPoStCompound<H>
+{
+    fn cache_prefix() -> String {
+        format!("proof-of-spacetime-fallback-{}", H::name())
+    }
+}
+
+impl<'a, H> CompoundProof<'a, Bls12, FallbackPoSt<'a, H>, FallbackPoStCircuit<'a, Bls12, H>>
+    for FallbackPoStCompound<H>
+where
+    H: 'a + Hasher,
+{
+    fn generate_public_inputs(
+        pub_inputs: &<FallbackPoSt<'a, H> as ProofScheme<'a>>::PublicInputs,
+        pub_params: &<FallbackPoSt<'a, H> as ProofScheme<'a>>::PublicParams,
+        partition_k: Option<usize>,
+    ) -> Result<Vec<Fr>> {
+        let mut inputs = Vec::new();
+
+        let por_pub_params = por::PublicParams {
+            leaves: (pub_params.sector_size as usize / NODE_SIZE),
+            private: true,
+        };
+
+        let num_sectors_per_chunk = pub_params.sector_count;
+
+        let partition_index = partition_k.unwrap_or(0);
+
+        let sectors = pub_inputs
+            .sectors
+            .chunks(num_sectors_per_chunk)
+            .nth(partition_index)
+            .ok_or_else(|| anyhow!("invalid number of sectors/partition index"))?;
+
+        for (i, sector) in sectors.iter().enumerate() {
+            // 1. Inputs for verifying comm_r = H(comm_c || comm_r_last)
+            inputs.push(sector.comm_r.into());
+
+            // 2. Inputs for verifying inclusion paths
+            for n in 0..pub_params.challenge_count {
+                let challenge_index = (partition_index * pub_params.sector_count
+                    + i * pub_params.challenge_count
+                    + n) as u64;
```

Comment:
I'm not sure this is right. Shouldn't it be `(partition_index * pub_params.sector_count + i) * pub_params.challenge_count + n`? Maybe I got that wrong, or maybe I'm missing something else — so let me also describe what I think you're trying to do. I think you're trying to assign a unique, sequential index to every challenge — starting from the first sector of the first partition, incrementing through all the challenges of that sector, and proceeding through the remaining sectors, then doing the same for each partition.

If that sounds right, what you have above seems off. I think the above will duplicate `challenge_index`es.

Reply:
cood catch, fixing

Reply:
fixed